### PR TITLE
Fix bug reporting out of order iteration

### PR DESCRIPTION
**Pull request by jkeiser**
Right now, loops aren't reporting the out of order iteration error when you actually descend into the loop. This masked a bug where we were reporting out of order iteration error for empty arrays! This fixes both bugs.

It has no impact on performance, as far as I can tell.

This PR is based against #1432, to avoid conflicts.

### CODE REVIEW

**Comment by lemire** [comment]
We are getting this error...

```

/drone/src/include/simdjson/generic/ondemand/value_iterator-inl.h:405:39: error: assuming signed overflow does not occur when assuming that (X + c) >= X is always true [-Werror=strict-overflow]
--
1855 | return _json_iter->depth() >= depth();

```

**Comment by jkeiser** [comment]
That error is so ridiculously annoying. It's never an actual error--it's generally because we're carefully designed so the optimizer can make assumptions like that!

**Comment by jkeiser** [comment]
OK, that should be fixed. We'll see obviously :)

File: include/simdjson/generic/ondemand/value_iterator-inl.h [hunk 1]
```diff
@@ -504,6 +509,10 @@ simdjson_really_inline bool value_iterator::is_at_start() const noexcept {
 simdjson_really_inline bool value_iterator::is_at_container_start() const noexcept {
   return _json_iter->token.index == _start_position + 1;
 }
+simdjson_really_inline bool value_iterator::is_at_iterator_start() const noexcept {
+  auto delta = _json_iter->token.index - _start_position;
+  return delta == 1 || delta == 2;
```
Comment (lemire) [inline include/simdjson/generic/ondemand/value_iterator-inl.h:514]: A comment here might be helpful to improve code legibility.

**Review by lemire — APPROVED** [review]

File: include/simdjson/generic/ondemand/value_iterator-inl.h (hunk 1 above)
Comment (jkeiser) [reply include/simdjson/generic/ondemand/value_iterator-inl.h:515]: :+1:

**Pull request merged**